cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/HTableNameConverter.java

```diff
@@ -16,18 +16,113 @@
+
+  private static String getHBaseTableName(String tableName) {
+    return encodeTableName(tableName);
+  }
+
+  private static String encodeTableName(String tableName) {
+    try {
+      return URLEncoder.encode(tableName, "ASCII");
+    } catch (UnsupportedEncodingException e) {
+      // this can never happen - we know that ASCII is a supported character set!
+      throw new RuntimeException(e);
+    }
+  }
+
+  private static String getBackwardCompatibleTableName(String tablePrefix, TableId tableId) {
+    String tableName = tableId.getTableName();
+    // handle table names in default namespace so we do not have to worry about upgrades
+    if (Constants.DEFAULT_NAMESPACE_ID.equals(tableId.getNamespace())) {
+      // if the table name starts with 'system.', then its a queue or stream table. Do not add namespace to table name
+      // e.g. namespace = default, tableName = system.queue.config. Resulting table name = cdap.system.queue.config
+      // also no need to prepend the table name if it already starts with 'user'.
+      // TODO: the 'user' should be prepended by the HBaseTableAdmin.
+      if (tableName.startsWith(String.format("%s.", Constants.SYSTEM_NAMESPACE)) ||
+          tableName.startsWith("user.")) {
+        return Joiner.on(".").join(tablePrefix, tableName);
+      }
+      // if the table name does not start with 'system.', then its a user dataset. Add 'user' to the table name to
+      // maintain backward compatibility. Also, do not add namespace to the table name
+      // e.g. namespace = default, tableName = purchases. Resulting table name = cdap.user.purchases
+      return Joiner.on(".").join(tablePrefix, "user", tableName);
+    }
+    // if the namespace is not default, do not need to change anything
+    return tableName;
+  }
+
+  /**
+   * @return Backward compatible, ASCII encoded table name
+   */
+  protected static String getHBaseTableName(CConfiguration cConf, TableId tableId) {
+    String tablePrefix = cConf.get(Constants.Dataset.TABLE_PREFIX);
+    Preconditions.checkArgument(tablePrefix != null, "Table prefix should not be null.");
+    return getHBaseTableName(getBackwardCompatibleTableName(tablePrefix, tableId));
+  }
+
   /**
    * Gets the system configuration table prefix
-   * @param tableName Full table name.
+   * @param hTableName Full HBase table name.
    * @return System configuration table prefix (full table name minus the table qualifier).
    * Example input: "cdap_ns.table.name"  -->  output: "cdap_system."   (hbase 94)
    * Example input: "cdap.table.name"     -->  output: "cdap_system."   (hbase 94. input table is in default namespace)
    * Example input: "cdap_ns:table.name"  -->  output: "cdap_system:"   (hbase 96, 98)
    */
-  public abstract String getSysConfigTablePrefix(String tableName);
+  public abstract String getSysConfigTablePrefix(String hTableName);
+
+  protected abstract TableId fromTableName(String hTableName);
```

Comment:
Why have this? I don't think we should have this. The signature is different for different HBase version, right? And it is only used within the HBase compat module, right?